### ISSUE TITLE
Understanding losses, while finetuning Gliner

### ISSUE THREAD

**Issue by dcrey7** [urchade/GLiNER#279]
Hi @urchade,

Thank you for providing all the examples and all the additional information about gliner in the repo. I tried finetuning gliner model using your example notebook with the focal loss parameters and without the focal loss parameters. I would like to know 
1) when using focal loss where is huge difference between training and validation loss? why does this happen and the results (F1 is also quite low)
2) when we dont use focal loss both training and validation loss are similar but they do not ever come down or go lower to 0?

your thoughts and insight are very valuable, please do let me know!

I have shared the script for the classification report as well

here are the results without focal loss parameters

<img width="443" height="725" alt="Image" src="https://github.com/user-attachments/assets/e98c2086-401b-4266-bbbc-b38c6e98687b" />

<img width="166" height="88" alt="Image" src="https://github.com/user-attachments/assets/8eb23c01-fd7c-49d8-98b6-399a796a67d7" />

and here are the results with focal loss parameters

<img width="675" height="788" alt="Image" src="https://github.com/user-attachments/assets/59d4a878-20fe-44a6-b109-9b2ad98d5e87" />

<img width="158" height="102" alt="Image" src="https://github.com/user-attachments/assets/86502b89-971f-4d4f-b64f-68128c351cc7" />

---

from sklearn.metrics import classification_report, precision_recall_fscore_support, accuracy_score

y_true, y_pred = [], []

for example in test_dataset:
    tokens = example['tokenized_text']
    text = " ".join(tokens)
    true_entities = example['ner']

    print("---------------")
    print("Text:", text)
    print("True entities:", true_entities)

    # Convert token indices to char indices (ground truth)
    token_char_starts = []
    char_pos = 0
    for token in tokens:
        token_char_starts.append(char_pos)
        char_pos += len(token) + 1  # +1 for the space

    true_spans = set()
    for start_token, end_token, label in true_entities:
        if end_token < len(tokens):
            char_start = token_char_starts[start_token]
            char_end = token_char_starts[end_token] + len(tokens[end_token])
            true_spans.add((char_start, char_end, label))

    # Predicted spans
    preds = trained_model.predict_entities(text, labels, threshold=0.65)
    print("PREDiction:",preds)
    pred_spans = set((pred['start'], pred['end'], pred['label']) for pred in preds)

    print("pred_spans:", pred_spans)
    print("true_spans:", true_spans)
    # Collect predictions and truths
    all_spans = true_spans.union(pred_spans)
    #print(all_spans)
    for span in all_spans:
        y_true.append(span[2] if span in true_spans else 'O')
        y_pred.append(span[2] if span in pred_spans else 'O')
    # Collect predictions and truths without 'O'
    #for span in pred_spans:
    #   y_pred.append(span[2])
    #for span in true_spans:
    #    y_true.append(span[2])
import pandas as pd
# Classification Report
print("Entity-level Classification Report:")

print(classification_report(y_true, y_pred, zero_division=0))

# Assuming y_true and y_pred are your label arrays
report = classification_report(y_true, y_pred, zero_division=0, output_dict=True)

# Convert to DataFrame
df_report = pd.DataFrame(report).transpose()

# Remove 'O' entity
df_report_filtered = df_report.drop('O', axis=0)

# Display updated classification report
#print(df_report_filtered)

precision, recall, f1, _ = precision_recall_fscore_support(y_true, y_pred, average='weighted', zero_division=0)
accuracy = accuracy_score(y_true, y_pred)

print(f"Precision: {precision:.4f}")
print(f"Recall:    {recall:.4f}")
print(f"F1-score:  {f1:.4f}")
print(f"Accuracy:  {accuracy:.4f}")
